Integrate support for building JavaScript modules within the build system

cmacq2 opened the pull request:
Integrate support for building JavaScript modules within the build system

Building JavaScript modules is opt-in, i.e. disabled by default.
Various 'flavours' of JavaScript module syntax are supported:

 - ECMAScript (es)
 - CommonJS (cjs)
 - Immediately Invoked Function Expression (iife)
 - Universal Module Definition (umd)

To enable a particular flavour one of the `--enable-javascript-*-module` flags may be passed with an argument of `yes` during `./configure`:

 - `--enable-javascript-es-module`
 - `--enable-javascript-common-js-module`
 - `--enable-javascript-iife-module`
 - `--enable-javascript-umd-module`

Additionally, `configure.ac` is further extended to look for an available version of `npm` if any of the supported module flavours are enabled. If no `npm` can be found and a module flavour is enabled, `./configure` will now fail with an error message accordingly. *Nota bene*: this check for `npm` applies if and only if modules are actually enabled, therefore by default Janus will still build without `npm`.

When modules are enabled, generated modules will be built in `./npm/bundles` (during `make`) and installed to `$PREFIX/share/janus/javascript` (during `make install`). Additionally `janus.js` itself is now always installed to `$PREFIX/share/janus/javascript` as well (even if no modules are enabled). The idea is that `$PREFIX/share/janus/javascript` acts as the installation location for the generically re-usable JavaScript libraries which are part of Janus.

cmacq2 commented:
This pull request effectively also obviates PR #792: the UMD wrapper is now auto-generated and therefore no longer require changes to `janus.js` itself.

thehunmonkgroup commented:
I support this approach over my original proposal in #792, it accomplishes the goal in a smart location with more flexible options.

I'm wondering about support for custom install locations -- both for a custom install of the npm executable, and the install location of the wrapped .js files. Neither of these is a deal-killer IMO, but worth considering.

Also, for those of us that use janus.js on a different server than the Janus process itself, it would be nice if there was a way to build just the wrapped javascript instead of having to build all of Janus to get that.

cmacq2 commented:
> both for a custom install of the npm executable

In `configure.ac` the PR includes this snippet:

```
AC_ARG_VAR(NPM,"npm executable to use")
AC_PATH_PROG(NPM,npm,,)
```
The purpose of that bit is:

 - To search the PATH for available `npm`
 - And allow people to pass in custom `npm` via the `NPM` variable: `./configure NPM=/npm/in/strange-location`. See: https://www.gnu.org/software/autoconf/manual/autoconf-2.64/html_node/Setting-Output-Variables.html

thehunmonkgroup commented:
Thanks for that clarification on the npm path, good to know!

cmacq2 commented:
> Also, for those of us that use janus.js on a different server than the Janus process itself, it would be nice if there was a way to build just the wrapped javascript instead of having to build all of Janus to get that.

As things stand with this PR, you could manually invoke `rollup` to do so:

```
# build just the ECMAScript flavour
cd npm && mkdir -p bundles && npm run rollup -- --f es --o bundles/janus.es.js
```

We could add a custom `make` target to automate that, but:

 - You'd still need to get past `./configure` first. Which, as things stand, assumes Janus is going to built, and therefore requires a complete Janus Gateway build environment.
 - It feels awkward to be using autotools configure+make in a 'pure' JavaScript project setup. You probably don't even want to be using autotools, likely you'd prefer to use plain npm + rollup.

So the question is if the convenience `make` target is worth it, even if it doesn't actually address your use case very well.

cmacq2 commented:
> and the install location of the wrapped .js files

What do you have in mind? You can already influence the installation directory of `make install` using the `--prefix` option during `./configure`.

thehunmonkgroup commented:
@cmacq2 you're right about ```--prefix```, I think that's enough.

Yeah, providing that as a make target feels ugly. What about providing a helper script that wraps all the commands? Or, at the very least, adding something in the documentation?

lminiero commented on Makefile.am:
```diff
@@ -439,6 +442,37 @@ SUBDIRS += docs
+	cd npm && $(NPM) run rollup -- --o $(patsubst npm/%,%,$@) --f $*
+
+endif
+##
```
Missing empty line here.

lminiero commented on configure.ac:
```diff
@@ -674,6 +722,23 @@ echo "Event handlers:"
 	[echo "    Sample event handler:  yes"],
 	[echo "    Sample event handler:  no"])
+AM_COND_IF([ENABLE_JAVASCRIPT_MODULES], [
+	echo "Using npm:                 $NPM"
```
Please put this indented in the JavaScript modules thingie: I don't want people to think you need npm to use Janus.

lminiero commented on npm/package.json:
```diff
@@ -0,0 +1,14 @@
+{
+  "name": "janus",
+  "description": "The Janus WebRTC Gateway",
```
It's "Server" now, makes people less confused about what we do :smile:

cmacq2 commented:
@lminiero I've incorporated your feedback, and rebased the PR on top of latest master.

lminiero commented:
Thanks as usual for the quick changes! I'll get back to you early next week with feedback, so that we can merge ASAP :+1:

cmacq2 commented:
> Yeah, providing that as a make target feels ugly. What about providing a helper script that wraps all the commands? Or, at the very least, adding something in the documentation?

Documentation: working on some. ;) It occurred to me the getting started docs should probably have an additional page on these module variants + a few pointers.

lminiero commented:
@atoppi can you review this when you have some time?

atoppi commented:
I've got some issues trying to compile JS with:

`./configure --prefix=/opt/janus --enable-javascript-es-module=yes && make && make install`

gives me the following error
```
cd npm && /home/atoppi/.nvm/versions/node/v6.11.4/bin/npm install
/usr/bin/env: ‘node’: No such file or directory
Makefile:2772: recipe for target 'npm/node_modules/rollup/bin/rollup' failed
make[2]: *** [npm/node_modules/rollup/bin/rollup] Error 127
```
I had to create a link with

`sudo ln -s /home/atoppi/.nvm/versions/node/v6.11.4/bin/node /usr/bin/node`

But the bundle js seems not correctly generated
```
atoppi@atoppi-Inspiron-7559:~/janus-gateway/npm/bundles$ cat janus.es.js
/*
 * Module shim for rollup.js to work with.
 * Simply re-export Janus from janus.js, the real 'magic' is in the rollup config
 */

/*
	The MIT License (MIT)

	Copyright (c) 2016 Meetecho

	Permission is hereby granted, free of charge, to any person obtaining
	a copy of this software and associated documentation files (the "Software"),
	to deal in the Software without restriction, including without limitation
	the rights to use, copy, modify, merge, publish, distribute, sublicense,
	and/or sell copies of the Software, and to permit persons to whom the
	Software is furnished to do so, subject to the following conditions:

	The above copyright notice and this permission notice shall be included
	in all copies or substantial portions of the Software.

	THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
	OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
	FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
	THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR
	OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
	ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
	OTHER DEALINGS IN THE SOFTWARE.
 */

// List of sessions

export default JanusOOPS;

```

cmacq2 commented:
@atoppi 
Looks like a partial add/commit on my side, that should be fixed now. Also rebased on top of latest master.
Question: did you make sure to run `./autogen.sh` prior to `./configure` ?

cmacq2 commented:
Also: judging by the diff the github PR system is borked: the module.js file should read `export Janus;` as does the branch in my repo: https://github.com/cmacq2/janus-gateway/blob/build-js-modules/npm/module.js

Maybe you can try pulling from there?

atoppi commented:
`atoppi@atoppi-Inspiron-7559:~/janus-gateway$ ./autogen.sh && ./configure NPM=/home/atoppi/.nvm/versions/node/v6.11.4/bin/npm --prefix=/opt/janus --enable-javascript-es-module=yes && make && sudo make install`

unfortunately this still fails
```
...
JavaScript modules:        yes
    Using npm:             /home/atoppi/.nvm/versions/node/v6.11.4/bin/npm
    ES syntax:             yes
    IIFE syntax:           no
    UMD syntax:            no
    CommonJS syntax:       no

If this configuration is ok for you, do a 'make' to start building Janus. A 'make install' will install Janus and its plugins to the specified prefix. Finally, a 'make configs' will install some sample configuration files too (something you'll only want to do the first time, though).

`which git` rev-parse HEAD | awk 'BEGIN {print "#include \"version.h\""} {print "const char *janus_build_git_sha = \"" $0"\";"} END {}' > version.c
echo "Wed Oct 18 10:48:19 CEST 2017" | awk 'BEGIN {} {print "const char *janus_build_git_time = \""$0"\";"} END {} ' >> version.c
make  all-recursive
make[1]: Entering directory '/home/atoppi/janus-gateway'
Making all in html
make[2]: Entering directory '/home/atoppi/janus-gateway/html'
make[2]: Nothing to be done for 'all'.
make[2]: Leaving directory '/home/atoppi/janus-gateway/html'
make[2]: Entering directory '/home/atoppi/janus-gateway'
`which git` rev-parse HEAD | awk 'BEGIN {print "#include \"version.h\""} {print "const char *janus_build_git_sha = \"" $0"\";"} END {}' > version.c
echo "Wed Oct 18 10:48:19 CEST 2017" | awk 'BEGIN {} {print "const char *janus_build_git_time = \""$0"\";"} END {} ' >> version.c
  CC       janus-version.o
  CCLD     janus
cd npm && /home/atoppi/.nvm/versions/node/v6.11.4/bin/npm install
/usr/bin/env: ‘node’: No such file or directory
Makefile:2772: recipe for target 'npm/node_modules/rollup/bin/rollup' failed
make[2]: *** [npm/node_modules/rollup/bin/rollup] Error 127
make[2]: Leaving directory '/home/atoppi/janus-gateway'
Makefile:2275: recipe for target 'all-recursive' failed
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory '/home/atoppi/janus-gateway'
Makefile:1049: recipe for target 'all' failed
make: *** [all] Error 2
```

atoppi commented:
Beside the current issue, I think that the `clean` target should remove the folders npm/bundles and npm/node_modules and never fail if npm is not  found.

cmacq2 commented:
Let's first fix the `make clean`, that should be most straightforward. The second invocation `npm install` can be worked around via `touch` in the `Makefile` rules.

cmacq2 commented:
- I've now fixed `make clean` to remove the generated bundles by registering them with `CLEANFILES`.
- I updated the npm install rule to `touch` the rollup script, so it isn't executed twice.

cmacq2 commented:
@atoppi I've also been playing with my own `PATH` to deliberately exclude `node` from it, like so:

```
which npm
which node
cd $HOME && ln -s "$(which npm)"
cd $OLDPWD
./autogen.sh
./configure  NPM=$HOME/npm --enable-javascript-es-module=yes --enable-javascript-iife-module=yes --enable-javascript-umd-module=yes --enable-javascript-common-js-module=yes --prefix=/home/prometheus/projects/janus/dist --enable-docs
PATH=/usr/bin:/bin make
```
And I can't provoke the `/usr/bin/env: ‘node’: No such file or directory` error. Also, when running through `npm` that shouldn't be happening if I understand the [docs](https://docs.npmjs.com/cli/run-script) correctly, `npm` should execute scripts with an adjusted `PATH` to include the `node` interpreter. What is your version of `npm`?

Mine is:
```
npm --version
4.1.2
```

According to the docs you could try adding `--scripts-prepend-node-path` to the `Makefile.am` rule that invokes `npm run`.

atoppi commented:
@cmacq2 I have npm version 3.10.10, that should be the one bundled inside the last LTS version of nodejs (my current nodejs is v6.11.4).
The installation has been accomplished with the popular `nvm` script (this easily permits multiple nodejs installations and the switch between the different versions).

Anyway all is working after last fix.

cmacq2 commented:
> Anyway all is working after last fix.

Ok, that's great to see. Do you think the PR can be merged as is, now?

lminiero commented:
I'll have to do a quick review myself too, and some rounds of compilation just to make sure it doesn't break anything for those who don't care about npm. I'll try to do that later today.

atoppi commented:
@cmacq2 what about the empty folder `bundles` that is being created in the source root (in my case `~/janus-gateway`) after `make`?

cmacq2 commented:
@atoppi fixed now, the bundles directory should no longer be created.

atoppi commented:
@cmacq2 ok working as expected.

cmacq2 commented:
@lminiero did you manage to find some time for your review?

lminiero commented:
@cmacq2 apologiesagain for the delay, just tried and I can confirm it seems to work beautifully, thanks! 

The configure flags could probably benefit from a "enable all JS modules" kind of thing like we added for plugins and transports, and I spotted a couple of small issues in the generated docs, but nothing I can't fix myself after merging. Another thing I noticed, though, is that it seems the npm folder is not added to the archive when you do a `make dist`, which basically builds a `.tar.gz` of the code. That said, I'm not sure anybody actually uses this at all, so we can ignore this for now and I can check how to fix that later on.

Merging, thanks again for the effort!

lminiero commented:
Updated the docs online after the few doc fixes (mainly making the table bootstrap-compliant, and double-dashes being collapsed to a single long dash in html): https://janus.conf.meetecho.com/docs/js-modules

lminiero commented:
Fixed the broken `make dist` and added a new `--enable-all-js-modules` configure flag too.

thehunmonkgroup commented:
@cmacq2 I dug through the doc of rollup.js, but wasn't able to figure this out:

I want to add adapter.js as a dependency in the define of the AMD wrapper, like this:
```javascript
define(['adapter'], function(adapter) {
});
```
But no matter what I've tried for command line switches to rollup, it just spits out an empty define wrapper, like this:
```javascript
define(function() {
});
```
Is there any way to get what I'm looking for?

cmacq2 commented:
@thehunmonkgroup 

If you want to `import` other modules, then you should `import` the modules directly in the JavaScript code. So you would need to create a new `module.js` entry point like this:

```
import adapter from 'webrtc-adapter';
// other imports elided

@JANUS_CODE@

export default Janus;
```

That said, `janus.js` has the `use***Dependencies()` methods to pass in dependencies, and if you don't pass in dependencies explicitly the defaults will be used and those may not match your `import` statements. Effectively, (for the most part) `janus.js` no longer has static dependencies like that, it's all deferred to the `Janus.init()` call. This also means that the `import` statements only make sense if these happen to align with the exact same defaults used in `Janus.useDefaultDependencies()` (e.g. 'adapter' from webrtc-adapter).

If not, you can either rewrite the `export` part to shim the `Janus.init` function or use the existing module as-is but write a shim module (e.g. `janus-integration`) on top which does much the same:

```
import Janus from './janus.es.js';
import adapter from 'webrtc-adapter';
// other imports elided

const adaptedJanus = function() {
  return Janus.apply(this, arguments);
};

adaptedJanus.init = options => {
   const dependencies = /* implement those here */;
   return Janus.init(Object.assign(options || {}, {dependencies}));
};

export default adaptedJanus;
```
And then point `rollup` to that module as the entry point instead or inject it directly inside your projects which are meant to consume the Janus library.

thehunmonkgroup commented:
@cmacq2 I'm sure I can figure out how to shim things in require.js, I was hoping to avoid the complexity :)

It seems like ```Janus.useDefaultDependencies()``` was written to allow overriding the defaults with custom deps: https://github.com/meetecho/janus-gateway/blob/master/html/janus.js#L47

But ```Janus.init()``` provides no way to leverage that: https://github.com/meetecho/janus-gateway/blob/master/html/janus.js#L202

And the way ```Janus.useDefaultDependencies()``` is written, it doesn't seem like it can be called outside of ```Janus.init()``` and have any effect.

Since I already have an ```adapter``` local in my require.js callback, I *could* pass that in the dependencies arg for ```Janus.init()```, but it seems like if I did that, the way it's written would require me to provide *all* the custom deps. Ideally, I could pass ```adapter``` into ```Janus.init()```, and since no other custom deps are supplied, the defaults would be used.

cmacq2 commented:
> It seems like Janus.useDefaultDependencies() was written to allow overriding the defaults with custom deps: 

Correct.

> And the way Janus.useDefaultDependencies() is written, it doesn't seem like it can be called outside of Janus.init() and have any effect.

Of course it can. You first create a `dependencies` object using one of those `use***Dependencies()` methods, then you pass it in to `Janus.init()` as part of the `options` argument. 

So your particular example could be expressed like this:

```
// construct your custom adapter
const myAdapter = ...;

// construct Janus deps: use your custom adapter, and defaults for everything else
const dependencies = Janus.useDefaultDependencies({adapter: myAdapter}); 

// finally, init:
Janus.init({dependencies});
```

thehunmonkgroup commented:
Yep, that workflow hadn't occurred to me. Thanks for pointing it out!

fbernhard commented:
Even though janus.js is licensed under the MIT license the package.json defines the license as GPL-3.0. Im my opinion the license property in the package.json only refers to the Javascript files in the NPM package and not Janus itself. With "npm install" I only install the Javascript code.

Can we change the license attribute in the package.json to MIT?

lminiero commented:
As I was saying on the group, it would be fine by me, but we'd have to change the [description](https://github.com/meetecho/janus-gateway/blob/master/npm/package.json#L3) to clarify the `package.json` file is for the JS hooks and not Janus itself. Suggestions on what to put there are welcome.

fbernhard commented:
Suggestion:

`"description": "The Janus WebRTC Server Javascript Module. IMPORTANT NOTE: the MIT License applies only to the Javascript files contained in this NPM package and NOT to Janus itself."`
